File: pascal/grep_std.pas
// fpgrep_std - idiomatic single-threaded Free Pascal grep -F clone.
{$mode objfpc}{$H+}
program grep_std;

uses
  BaseUnix, SysUtils;

var
  Pat: RawByteString = '';
  LPat: RawByteString = '';
  PatSet: Boolean = False;
  CI: Boolean = False;
  Recursive: Boolean = False;
  Multi: Boolean = False;
  Matched: Boolean = False;
  // manual output buffer flushed to fd 1
  OutBuf: array of Byte;
  OutLen: SizeInt = 0;

const
  OUTCAP = 1 shl 16;

procedure FlushOut;
begin
  if OutLen > 0 then
  begin
    FpWrite(1, OutBuf[0], OutLen);
    OutLen := 0;
  end;
end;

procedure OutBytes(const P: PByte; N: SizeInt);
begin
  if N <= 0 then Exit;
  if N >= OUTCAP then
  begin
    FlushOut;
    FpWrite(1, P^, N);
    Exit;
  end;
  if OutLen + N > OUTCAP then FlushOut;
  Move(P^, OutBuf[OutLen], N);
  Inc(OutLen, N);
end;

procedure OutByte(B: Byte);
begin
  if OutLen + 1 > OUTCAP then FlushOut;
  OutBuf[OutLen] := B;
  Inc(OutLen);
end;

procedure OutStr(const S: RawByteString);
begin
  if Length(S) > 0 then
    OutBytes(PByte(@S[1]), Length(S));
end;

// ASCII-only length-preserving lowercase (matches grep -iF).
procedure AsciiLower(Dst, Src: PByte; N: SizeInt);
var
  i: SizeInt;
  b: Byte;
begin
  for i := 0 to N - 1 do
  begin
    b := Src[i];
    if (b >= 65) and (b <= 90) then
      Dst[i] := b + 32
    else
      Dst[i] := b;
  end;
end;

// Hand-rolled byte scan: first index >= start where needle occurs in hay[0..hlen-1].
// Returns -1 if not found. Empty needle returns start.
function ByteIndex(Hay: PByte; HLen: SizeInt; Needle: PByte; NLen: SizeInt; Start: SizeInt): SizeInt;
var
  i, j: SizeInt;
  first: Byte;
  ok: Boolean;
begin
  if NLen = 0 then
  begin
    ByteIndex := Start;
    Exit;
  end;
  if Start < 0 then Start := 0;
  first := Needle[0];
  i := Start;
  while i + NLen <= HLen do
  begin
    if Hay[i] = first then
    begin
      ok := True;
      j := 1;
      while j < NLen do
      begin
        if Hay[i + j] <> Needle[j] then begin ok := False; Break; end;
        Inc(j);
      end;
      if ok then begin ByteIndex := i; Exit; end;
    end;
    Inc(i);
  end;
  ByteIndex := -1;
end;

// last newline (10) before index m in data[0..m-1], +1; or 0.
function LineStart(Data: PByte; M: SizeInt): SizeInt;
var
  i: SizeInt;
begin
  i := M - 1;
  while i >= 0 do
  begin
    if Data[i] = 10 then begin LineStart := i + 1; Exit; end;
    Dec(i);
  end;
  LineStart := 0;
end;

// first newline (10) at/after m; or len.
function LineEnd(Data: PByte; M, Len: SizeInt): SizeInt;
var
  i: SizeInt;
begin
  i := M;
  while i < Len do
  begin
    if Data[i] = 10 then begin LineEnd := i; Exit; end;
    Inc(i);
  end;
  LineEnd := Len;
end;

procedure SearchFile(const Path: RawByteString);
var
  Data: RawByteString;
  Low: RawByteString;
  Hay, Needle: PByte;
  Len, NLen, Peek, Pos, m, ls, le, i, got, off: SizeInt;
  fd: cint;
  st: Stat;
begin
  fd := FpOpen(Path, O_RDONLY);
  if fd < 0 then Exit;
  if FpFStat(fd, st) <> 0 then begin FpClose(fd); Exit; end;
  Len := st.st_size;
  SetLength(Data, Len);
  off := 0;
  while off < Len do
  begin
    got := FpRead(fd, Data[off + 1], Len - off);
    if got <= 0 then Break;
    Inc(off, got);
  end;
  FpClose(fd);
  if off < Len then Len := off;
  SetLength(Data, Len);

  Peek := Len;
  if Peek > 65536 then Peek := 65536;
  for i := 0 to Peek - 1 do
    if Byte(Data[i + 1]) = 0 then Exit; // binary

  if Len = 0 then
    Hay := nil
  else
    Hay := PByte(@Data[1]);

  if CI then
  begin
    SetLength(Low, Len);
    if Len > 0 then
      AsciiLower(PByte(@Low[1]), Hay, Len);
    if Len > 0 then Hay := PByte(@Low[1]) else Hay := nil;
    Needle := PByte(LPat);
    NLen := Length(LPat);
  end
  else
  begin
    Needle := PByte(Pat);
    NLen := Length(Pat);
  end;

  Pos := 0;
  while Pos < Len do  // < not <= : empty-pattern fix
  begin
    m := ByteIndex(Hay, Len, Needle, NLen, Pos);
    if m < 0 then Break;
    ls := LineStart(PByte(@Data[1]), m);
    le := LineEnd(PByte(@Data[1]), m, Len);
    Matched := True;
    if Multi then
    begin
      OutStr(Path);
      OutByte(Ord(':'));
    end;
    if le > ls then
      OutBytes(@Data[ls + 1], le - ls);
    OutByte(10);
    Pos := le + 1;
  end;
end;

procedure WalkDir(const Dir: RawByteString); forward;

procedure Visit(const Path: RawByteString);
var
  st: Stat;
begin
  if FpLStat(Path, st) <> 0 then Exit;
  if fpS_ISLNK(st.st_mode) then Exit; // never follow symlinks
  if fpS_ISDIR(st.st_mode) then
  begin
    if Recursive then WalkDir(Path);
  end
  else if fpS_ISREG(st.st_mode) then
    SearchFile(Path);
end;

procedure WalkDir(const Dir: RawByteString);
var
  pd: PDir;
  ent: PDirent;
  name: RawByteString;
begin
  pd := FpOpenDir(Dir);
  if pd = nil then Exit;
  repeat
    ent := FpReadDir(pd^);
    if ent = nil then Break;
    name := PChar(@ent^.d_name[0]);
    if (name = '.') or (name = '..') then Continue;
    Visit(Dir + '/' + name);
  until False;
  FpCloseDir(pd^);
end;

procedure Usage;
const
  Msg: RawByteString = 'usage: fpgrep [-r] [-i] PATTERN PATH...'#10;
begin
  FpWrite(2, Msg[1], Length(Msg));
  Halt(2);
end;

var
  Paths: array of RawByteString;
  NPaths: SizeInt = 0;
  i, k: Integer;
  a: RawByteString;
  NoMore: Boolean = False;
  c: Char;
  st: Stat;
begin
  SetLength(OutBuf, OUTCAP);
  SetLength(Paths, ParamCount);
  for i := 1 to ParamCount do
  begin
    a := ParamStr(i);
    if (not NoMore) and (Length(a) >= 2) and (a[1] = '-') then
    begin
      if a = '--' then begin NoMore := True; Continue; end;
      for k := 2 to Length(a) do
      begin
        c := a[k];
        case c of
          'i': CI := True;
          'r': Recursive := True;
        else
          Usage;
        end;
      end;
    end
    else if not PatSet then
    begin
      Pat := a;
      PatSet := True;
    end
    else
    begin
      Paths[NPaths] := a;
      Inc(NPaths);
    end;
  end;

  if (not PatSet) or (NPaths = 0) then Usage;

  SetLength(LPat, Length(Pat));
  if Length(Pat) > 0 then
    AsciiLower(PByte(@LPat[1]), PByte(@Pat[1]), Length(Pat));
  Multi := Recursive or (NPaths > 1);

  for i := 0 to NPaths - 1 do
  begin
    if FpLStat(Paths[i], st) <> 0 then Continue;
    if fpS_ISDIR(st.st_mode) then
    begin
      if Recursive then WalkDir(Paths[i]);
    end
    else if fpS_ISREG(st.st_mode) then
      SearchFile(Paths[i])
    else if fpS_ISLNK(st.st_mode) then
    begin
      // top-level symlink: stat through it to decide file vs dir like grep does?
      // spec: use FpLStat to detect; skip symlinks during walk. Top-level arg:
      // grep -F follows top-level symlink args. Resolve once.
      if FpStat(Paths[i], st) = 0 then
      begin
        if fpS_ISREG(st.st_mode) then SearchFile(Paths[i])
        else if fpS_ISDIR(st.st_mode) and Recursive then WalkDir(Paths[i]);
      end;
    end;
  end;

  FlushOut;
  if Matched then Halt(0) else Halt(1);
end.

File: swift/grep_std.swift
// swiftgrep_std - idiomatic single-threaded Swift grep -F clone.
// Native LLVM-compiled (swiftc -O) with ARC (a third memory model: neither
// tracing-GC nor manual malloc/free). Recursive walk via Glibc opendir/readdir
// (skip symlinks, don't follow). Whole-file read. SCAN: byte-oriented via
// Glibc memmem() -- grep -F is a literal byte search, and memmem is the fast
// idiomatic option (Foundation Data.range(of:) would box bytes and is slower).
// ASCII-only -i folding (map A-Z, not Unicode). NUL-in-first-64KB binary skip.
// Mirrors c/grep_std.c semantics.
#if canImport(Glibc)
import Glibc
#endif

var pat: [UInt8] = []
var lpat: [UInt8] = []
var ci = false
var recursive = false
var multi = false
var matched = false

// reused output buffer flushed to fd 1
let OUTCAP = 1 << 16
var outBuf = [UInt8](repeating: 0, count: 1 << 16)
var outLen = 0

@inline(__always)
func flushOut() {
    if outLen > 0 {
        outBuf.withUnsafeBytes { _ = write(1, $0.baseAddress, outLen) }
        outLen = 0
    }
}

func outBytes(_ p: UnsafePointer<UInt8>, _ n: Int) {
    if n == 0 { return }
    if n >= OUTCAP {
        flushOut()
        _ = write(1, p, n)
        return
    }
    if outLen + n > OUTCAP { flushOut() }
    outBuf.withUnsafeMutableBufferPointer { dst in
        _ = memcpy(dst.baseAddress! + outLen, p, n)
    }
    outLen += n
}

@inline(__always)
func outByte(_ b: UInt8) {
    if outLen + 1 > OUTCAP { flushOut() }
    outBuf[outLen] = b
    outLen += 1
}

// ASCII-only, length-preserving lowercase (matches grep -iF).
func asciiLower(_ dst: inout [UInt8], _ src: UnsafePointer<UInt8>, _ n: Int) {
    dst.withUnsafeMutableBufferPointer { d in
        for i in 0..<n {
            let b = src[i]
            d[i] = (b >= 0x41 && b <= 0x5A) ? b &+ 32 : b
        }
    }
}

// First index >= from where needle occurs in hay[0..len]; -1 if none.
// Empty needle returns from. Uses memmem (Glibc) for the literal scan.
@inline(__always)
func byteIndex(_ hay: UnsafePointer<UInt8>, _ len: Int,
               _ needle: UnsafePointer<UInt8>, _ nlen: Int, _ from: Int) -> Int {
    if nlen == 0 { return from }
    if from + nlen > len { return -1 }
    guard let h = memmem(hay + from, len - from, needle, nlen) else { return -1 }
    return UnsafePointer(h.assumingMemoryBound(to: UInt8.self)) - hay
}

@inline(__always)
func lineStart(_ data: UnsafePointer<UInt8>, _ m: Int) -> Int {
    var i = m - 1
    while i >= 0 {
        if data[i] == 0x0A { return i + 1 }
        i -= 1
    }
    return 0
}

@inline(__always)
func lineEnd(_ data: UnsafePointer<UInt8>, _ m: Int, _ len: Int) -> Int {
    var i = m
    while i < len {
        if data[i] == 0x0A { return i }
        i += 1
    }
    return len
}

var lowbuf: [UInt8] = []  // reused ASCII-lowercase scratch

// Read an entire fd into a growable buffer.
func readAll(_ fd: Int32) -> [UInt8] {
    var buf = [UInt8](repeating: 0, count: 65536)
    var off = 0
    while true {
        if off == buf.count { buf.append(contentsOf: repeatElement(0, count: buf.count)) }
        let got = buf.withUnsafeMutableBufferPointer { b in
            read(fd, b.baseAddress! + off, b.count - off)
        }
        if got <= 0 { break }
        off += got
    }
    buf.removeLast(buf.count - off)
    return buf
}

func searchFile(_ path: String) {
    let fd = open(path, O_RDONLY)
    if fd < 0 { return }
    let data = readAll(fd)
    close(fd)
    let len = data.count
    if len == 0 { return }

    let peek = len < 65536 ? len : 65536
    let isBinary = data.withUnsafeBufferPointer { b -> Bool in
        memchr(b.baseAddress!, 0, peek) != nil
    }
    if isBinary { return }

    if ci {
        if lowbuf.count < len { lowbuf = [UInt8](repeating: 0, count: len) }
        data.withUnsafeBufferPointer { b in asciiLower(&lowbuf, b.baseAddress!, len) }
    }

    let pathBytes = Array(path.utf8)
    data.withUnsafeBufferPointer { db in
        let dptr = db.baseAddress!
        let hayArr = ci ? lowbuf : data
        hayArr.withUnsafeBufferPointer { hb in
            let hay = hb.baseAddress!
            let needle = ci ? lpat : pat
            needle.withUnsafeBufferPointer { nb in
                let nptr = nb.baseAddress!
                let nlen = needle.count
                var pos = 0
                // pos < len (not <=): empty needle would otherwise yield a phantom
                // match at pos==len; grep -F "" prints each line once.
                while pos < len {
                    let m = byteIndex(hay, len, nptr, nlen, pos)
                    if m < 0 { break }
                    let ls = lineStart(dptr, m)
                    let le = lineEnd(dptr, m, len)
                    matched = true
                    if multi {
                        pathBytes.withUnsafeBufferPointer { outBytes($0.baseAddress!, $0.count) }
                        outByte(0x3A)
                    }
                    if le > ls { outBytes(dptr + ls, le - ls) }
                    outByte(0x0A)
                    pos = le + 1
                }
            }
        }
    }
}

func walkDir(_ dir: String) {
    guard let pd = opendir(dir) else { return }
    while let ent = readdir(pd) {
        var nameBuf = ent.pointee.d_name
        let name = withUnsafePointer(to: &nameBuf) {
            $0.withMemoryRebound(to: CChar.self, capacity: 256) { String(cString: $0) }
        }
        if name == "." || name == ".." { continue }
        visit(dir + "/" + name)
    }
    closedir(pd)
}

func visit(_ path: String) {
    var st = stat()
    if lstat(path, &st) != 0 { return }
    let mode = st.st_mode & S_IFMT
    if mode == S_IFLNK { return }  // never follow symlinks during recursion
    if mode == S_IFDIR {
        if recursive { walkDir(path) }
    } else if mode == S_IFREG {
        searchFile(path)
    }
}

func usage() {
    let msg = "usage: swiftgrep_std [-r] [-i] PATTERN PATH...\n"
    let b = Array(msg.utf8)
    b.withUnsafeBufferPointer { _ = write(2, $0.baseAddress, $0.count) }
    exit(2)
}

func run() {
    var patStr = ""
    var patSet = false
    var paths: [String] = []
    var noMore = false
    let args = CommandLine.arguments
    for a in args[1...] {
        let ab = Array(a.utf8)
        if !noMore && ab.count >= 2 && ab[0] == 0x2D {
            if a == "--" { noMore = true; continue }
            for c in ab[1...] {
                if c == 0x69 { ci = true }          // 'i'
                else if c == 0x72 { recursive = true } // 'r'
                else { usage() }
            }
        } else if !patSet {
            patStr = a
            patSet = true
        } else {
            paths.append(a)
        }
    }
    if !patSet || paths.isEmpty { usage() }

    pat = Array(patStr.utf8)
    lpat = [UInt8](repeating: 0, count: pat.count)
    pat.withUnsafeBufferPointer { asciiLower(&lpat, $0.baseAddress!, pat.count) }
    multi = recursive || paths.count > 1

    for p in paths {
        var st = stat()
        if lstat(p, &st) != 0 { continue }
        let mode = st.st_mode & S_IFMT
        if mode == S_IFDIR {
            if recursive { walkDir(p) }
        } else if mode == S_IFREG {
            searchFile(p)
        } else if mode == S_IFLNK {
            // top-level symlink arg: resolve once like grep -F does.
            if stat(p, &st) == 0 {
                let m2 = st.st_mode & S_IFMT
                if m2 == S_IFREG { searchFile(p) }
                else if m2 == S_IFDIR && recursive { walkDir(p) }
            }
        }
    }

    flushOut()
    exit(matched ? 0 : 1)
}

run()
